### ISSUE TITLE
Variable is of type 'never' after potential assignment in a forEach lambda

### ISSUE THREAD

**Issue by billti** [microsoft/TypeScript#47399]
# Bug Report

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (including the nightly), and I reviewed the FAQ for entries about `never`.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/C4TwDgpgBAKgThCA1AlhA7gSWBAtlAXigG8oUATALigDsBXXAIwjgG4oBfAKC4DcBDOFGAII1eIlQZseANoBdQlFnEK1AIwcANCTVQA7B3ms+gqADMA9nRrkAqgGcW40VKw58AH1p0ANr6V6fxMuEUQAOis4AFF+AGMACwAKKCSIXzwASkIAPhIuKEKyc1T0vHCKKDyAVmyrG3snISIy3BMOTJCAei6oTBocfxQnGicyB2F03xQaAHMoXGgAcnrbRxYl8eFwZaDfTYSWCHCuFBKk1caWKABCAiI97OIeIqgewnuoaIAlb4B5b7UJYABTglkgcFAZHIUHIlggExolmAUAgAA9hijLDRtpBaBBeBsPgQCkUMljKkRLus4BVyO0uEA)

### 💻 Code

```ts
type TreeViewItem = { id: number; }

var tree: TreeViewItem[] = [{id: 1}, {id: 7}];
var foundUser: TreeViewItem | null = null;

tree.forEach( (elem) => {
    if (elem.id > 5) foundUser = elem;
});

// Intellisense is telling me 'foundUser' is type 'null' here.
if (foundUser !== null) {

    // === ERROR: 'Property id does not exist on type never' ===
    let oid = foundUser.id;
}

```

### 🙁 Actual behavior

See the error about `foundUser` being type `never` near the end. The type checker should detect the possible assignment inside the forEach lambda expression. (If the `forEach` check is replaced with something more inline, like `for (let x in tree) { if (tree[x].id > 5) foundUser = tree[x]; }`, then it behaves correctly)

### 🙂 Expected behavior

I expect that after the potential assignment inside the forEach lambda, the `foundUser` is of type `TreeViewItem | null` again.

This is a common pattern, and seems to have been present for a while, so I'm guessing it's a known issue that I just couldn't find. Just checking though!

**Comment by MartinJohns** [comment]
Another duplicate of #9998.

**Comment by billti** [comment]
Thanks @MartinJohns. I did browse the initial post on that issue (though not all the ~350 comments!) before logging, but this seemed a more straight-forward case I'm surprised isn't treated more optimistically, being there's a direct assignment to the variable in an expression in-between the declaration and the error site. Close as dupe if you feel it's well-trodden ground already. Thanks!

**Comment by MartinJohns** [comment]
The TypeScript team already marked the issue. In a few days the bot will automatically close the issue.

**Comment by MartinJohns** [comment]
@billti The closest solution to this "straight-forward case" would be #11498. But there hasn't been any movement there either.

A simple workaround is to avoid the narrowing upon assignment:

    const foundUser: TreeViewItem | null = null as TreeViewItem | null;
    // Now the type is "TreeViewItem | null", not "null".

**Comment by fatcerberus** [comment]
Once #46827 lands you'll also be able to do:
```ts
const foundUser = null satisfies TreeViewItem | null;
```
